frame/assets/src/lib.rs

```diff
@@ -1088,546 +1340,123 @@ impl<T: Config> Pallet<T> {
-
-			// Cannot create a new account with a balance that is below minimum...
-			assert_noop!(Assets::mint(Origin::signed(1), 0, 2, 9), Error::<Test>::BalanceLow);
-			assert_noop!(Assets::transfer(Origin::signed(1), 0, 2, 9), Error::<Test>::BalanceLow);
-			assert_noop!(Assets::force_transfer(Origin::signed(1), 0, 1, 2, 9), Error::<Test>::BalanceLow);
-
-			// When deducting from an account to below minimum, it should be reaped.
-
-			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 91));
-			assert!(Assets::balance(0, 1).is_zero());
-			assert_eq!(Assets::balance(0, 2), 100);
-			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 1);
-
-			assert_ok!(Assets::force_transfer(Origin::signed(1), 0, 2, 1, 91));
-			assert!(Assets::balance(0, 2).is_zero());
-			assert_eq!(Assets::balance(0, 1), 100);
-			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 1);
-
-			assert_ok!(Assets::burn(Origin::signed(1), 0, 1, 91));
-			assert!(Assets::balance(0, 1).is_zero());
-			assert_eq!(Asset::<Test>::get(0).unwrap().accounts, 0);
-		});
-	}
-
-	#[test]
-	fn querying_total_supply_should_work() {
-		new_test_ext().execute_with(|| {
-			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
-			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
-			assert_eq!(Assets::balance(0, 1), 100);
-			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 50));
-			assert_eq!(Assets::balance(0, 1), 50);
-			assert_eq!(Assets::balance(0, 2), 50);
-			assert_ok!(Assets::transfer(Origin::signed(2), 0, 3, 31));
-			assert_eq!(Assets::balance(0, 1), 50);
-			assert_eq!(Assets::balance(0, 2), 19);
-			assert_eq!(Assets::balance(0, 3), 31);
-			assert_ok!(Assets::burn(Origin::signed(1), 0, 3, u64::max_value()));
-			assert_eq!(Assets::total_supply(0), 69);
-		});
-	}
+	fn do_destroy(
+		id: T::AssetId,
+		witness: DestroyWitness,
+		maybe_check_owner: Option<T::AccountId>,
+	) -> DispatchResult {
+		Asset::<T>::try_mutate_exists(id, |maybe_details| {
+			let mut details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+			if let Some(check_owner) = maybe_check_owner {
+				ensure!(details.owner == check_owner, Error::<T>::NoPermission);
+			}
+			ensure!(details.accounts == witness.accounts, Error::<T>::BadWitness);
+			ensure!(details.sufficients == witness.sufficients, Error::<T>::BadWitness);
+			ensure!(details.approvals == witness.approvals, Error::<T>::BadWitness);
 
-	#[test]
-	fn transferring_amount_below_available_balance_should_work() {
-		new_test_ext().execute_with(|| {
-			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
-			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
-			assert_eq!(Assets::balance(0, 1), 100);
-			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 50));
-			assert_eq!(Assets::balance(0, 1), 50);
-			assert_eq!(Assets::balance(0, 2), 50);
-		});
-	}
+			for (who, v) in Account::<T>::drain_prefix(id) {
+				Self::dead_account(&who, &mut details, v.sufficient);
+			}
```

Comment:
It may be possible the witness data is too large to execute in one block, making it impossible to kill an asset.

Might be worth having some logic which:
* freezes the asset
* tries to kill as many as possible
* re-callable until everything is removed

Reply:
this would be managed by the asset admin off-chain. they already have the ability to freeze the asset and nullify accounts.